lib/active_merchant/billing/gateways/rapyd.rb

```diff
@@ -214,11 +218,29 @@ def add_payment_urls(post, options)
       end
 
       def add_customer_object(post, payment, options)
-        post[:name] = "#{payment.first_name} #{payment.last_name}" unless payment.is_a?(String)
         phone = options.dig(:billing_address, :phone) .gsub(/\D/, '') unless options[:billing_address].nil?
         post[:phone_number] = phone || options.dig(:customer, :phone_number)
         post[:email] = options[:email] || options.dig(:customer, :email)
-        post[:addresses] = options.dig(:customer, :addresses) if USA_PAYMENT_METHODS.include?(options[:pm_type])
+        post[:customer] = {}
+        post[:customer][:name] = "#{payment.first_name} #{payment.last_name}" unless payment.is_a?(String)
+        post[:customer][:addresses] = [address(options)]
+      end
+
+      def address(options)
+        return unless address = options[:billing_address]
+
+        formatted_address = {}
+
+        formatted_address[:name] = address[:name] if address[:name]
+        formatted_address[:line_1] = address[:address1] if address[:address1]
+        formatted_address[:line_2] = address[:address2] if address[:address2]
+        formatted_address[:city] = address[:city] if address[:city]
+        formatted_address[:state] = address[:state] if address[:state]
+        formatted_address[:country] = address[:country] if address[:country]
+        formatted_address[:zip] = address[:zip] if address[:zip]
+        formatted_address[:phone_number] = address[:phone].gsub(/\D/, '') if address[:phone]
+
+        formatted_address
       end
 
       def add_customer_id(post, options)
```

Comment:
`add_customer_id` sets `post[:customer] = options[:customer_id] if options[:customer_id]`. So in places where it's called after `add_customer_object` is called, like `add_auth_purchase`, it would currently override the customer object built in `add_customer_object`. Is that expected? If so, I would recommend adding a guard to `add_customer_object` to not bother building the customer object unless there's a `customer_id` option present.

Reply:
Thanks for calling this out! It seems that `customer` and `customer_id` usage has changed. We should not expect a `customer` in options but instead a possible `customer_id` in options.